tests/index.js

```diff
@@ -50,79 +74,131 @@ describe('broccoli-sane-watcher', function (done) {
 
     var builder = new broccoli.Builder(filter);
     watcher = new Watcher(builder);
-    watcher.on('change', function (results) {
-      assert.equal(results.directory, 'output');
+    watcher.on('change', eventHandler(function (results) {
+      expect(results.directory).to.equal('output');
+
       if (changes++) {
-        done();
+        defer.resolve();
       } else {
         fs.writeFileSync('tests/fixtures/a/file.js');
       }
-    });
+    }, defer));
+
+    watcher.on('error', errorHandler(defer));
+
+    return defer.promise;
+  });
+
+  it('cleans up heimdall nodes after builds and rebuilds', function() {
+    function countNodes() {
+      var count = 0;
 
-    watcher.on('error', function (error) {
-      assert.ok(false, error.message);
-      done();
+      heimdall.visitPreOrder(function () {
+        ++count;
+      });
+
+      return count;
+    }
+
+    var defer = RSVP.defer();
+
+    expect(countNodes()).to.equal(1, 'there is only the heimdall root node');
+
+    fs.mkdirSync('tests/fixtures/a');
+    var filter = new TestFilter(['tests/fixtures/a'], function () {
+      return 'output';
     });
+
+    var builder = new broccoli.Builder(filter);
+    watcher = new Watcher(builder);
+
+    watcher.on('error', errorHandler(defer));
+
+    var builds = 0;
+
+    watcher.on('change', eventHandler(function () {
+      // we build 3 times, but each time we only have the nodes from the current
+      // build (not previous builds)
+      expect(countNodes()).to.equal(3, 'we have expected number of nodes (root, testfilter, test/fixtures/a)');
+
+      var build = ++builds;
+      if (build === 1) {
+        fs.writeFileSync('tests/fixtures/a/foo.js');
+      } else if (build == 2) {
+        fs.writeFileSync('tests/fixtures/a/bar.js');
+      } else {
+        defer.resolve();
+      }
+    }, defer));
+
+    return defer.promise;
   });
 
-  it('should emit an error when a filter errors', function (done) {
+  it('should emit an error when a filter errors', function () {
     fs.mkdirSync('tests/fixtures/a');
     var filter = new TestFilter(['tests/fixtures/a'], function () {
       throw new Error('filter error');
     });
     var count = 0;
     var builder = new broccoli.Builder(filter);
     watcher = new Watcher(builder);
-    watcher.on('change', function (results) {
+
+    var defer = RSVP.defer();
+
+    watcher.on('change', eventHandler(function (results) {
       count++;
-      assert.equal(count, 2, "only the second build should be here");
-      assert.equal(results.directory, 'output');
-      done();
-    });
-    watcher.on('error', function (error) {
+      expect(count).to.equal(2, 'only the second build should be here');
+      expect(results.directory).to.equal('output');
```

Comment:
see [eventHandler](https://github.com/ember-cli/broccoli-sane-watcher/blob/f3794626c009a615e268a1971ab46d550602f8d4/tests/index.js#L33-L41).  The motivation of the changes to this test is precisely to prevent the problem of tests merely timing out on failures (which is what i see in master)